### PR TITLE
Add new test case for possible bug in BufReader

### PR DESCRIPTION
**Pull request by andre-vm**
When reading a large chunk from a BufReader, if all the bytes from the buffer have been already consumed, the internal buffer is bypassed entirely. However, it is not invalidated, and it's possible to access its contents using the `seek_relative` method, because it tries to reuse the existing buffer.

### CODE REVIEW

**Comment by joshtriplett** [comment]
I think the test case makes sense; if the buffer still exists, seeking to a location within it seems reasonable.

I don't think it makes sense to throw away the buffer until it gets overwritten. But once it *does* get overwritten, it needs invalidating.

Do you have a test case in which you can read data from the buffer that you shouldn't be able to?

**Review by joshtriplett — CHANGES_REQUESTED** [review]
I think the test case makes sense; if the buffer still exists, seeking to a location within it seems reasonable.

I don't think it makes sense to throw away the buffer until it gets overwritten. But once it *does* get overwritten, it needs invalidating.

Do you have a test case in which you can read data from the buffer that you shouldn't be able to? (Or am I misunderstanding what you're trying to demonstrate here?)

**Comment by Milack27** [comment]
@joshtriplett Yes, I believe the test case that I presented already shows that it is possible to access data from the buffer when we shouldn't:

1. `[5, 6, 7]` is read and stored in the buffer;
1. `[0, 1, 2, 3, 4]` is read bypassing the buffer. The buffer should now be invalidated but it's not, the chunk `[5, 6, 7]` remains there. At this point, we're at the end of the stream.
1. We call seek(-2). The buffer is 3 bytes long, so BufReader understands that we're still in its range, and that it doesn't need to fetch data from the inner reader.
1. We read again, but instead of getting `[3, 4]` the test fails and we get `[6, 7]`, which was incorrectly stored in the buffer.

**Comment by Milack27** [comment]
By the way, sorry for the mess. I'd like to squash those mistaken commits and force push them to keep it organized, but I don't have good internet access right now. I didn't even clone the Rust repository, I edited the file and commited on GitHub itself.

**Comment by joshtriplett** [comment]
On Mon, Mar 04, 2019 at 03:14:57PM -0800, André Vicente Milack wrote:
> @joshtriplett Yes, I believe the test case that I presented already shows that it is possible to access data from the buffer when we shouldn't:
> 
> 1. `[5, 6, 7]` is read and stored in the buffer;
> 1. `[0, 1, 2, 3, 4]` is read bypassing the buffer. The buffer should now be invalidated but it's not, the chunk `[5, 6, 7]` remains there. At this point, we're at the end of the stream.
> 1. We call seek(-2). The buffer is 3 bytes long, so BufReader understands that we're still in its range, and that it doesn't need to fetch data from the inner reader.
> 1. We read again, but instead of getting `[3, 4]` the test fails and we get `[6, 7]`, which was incorrectly stored in the buffer.

Ah, I see! And does the test pass if you uncomment the commented lines?

If so, please do go ahead and uncomment it and document the reason for
it, and I'd be happy to review that.

**Comment by Milack27** [comment]
Yes, the test passes when we uncomment those lines. I'll do it as soon as possible.

**Comment by Milack27** [comment]
@joshtriplett I realized that this problem not only happened in the `read` method, but also in the `seek` method. I just fixed the problem on both methods (I hope I didn't forget anything this time) and added another test case to cover `seek` as well.

**Comment by joshtriplett** [comment]
I feel like I need an extra pair of eyes here. I can see how this makes sense, but it feels like it might be invalidating more than it needs to, and I also wonder why it looks like seek's control flow can pass through both the invalidates you added in one call.

@Diggsey, @sfackler, it looks like you've done a fair bit of work in this area, can you take a look, please?

**Comment by Milack27** [comment]
> [...] and I also wonder why it looks like seek's control flow can pass through both the invalidates you added in one call.

I had the same question when I first read that code. It seems to me that the invalidation inside the else is there because the seek just below it can fail and return without passing through the invalidation outside the else.

File: src/libstd/io/buffered.rs [hunk 1]
```diff
@@ -323,14 +326,18 @@ impl<R: Seek> Seek for BufReader<R> {
             } else {
                 // seek backwards by our remainder, and then by the offset
                 self.inner.seek(SeekFrom::Current(-remainder))?;
-                self.pos = self.cap; // empty the buffer
+                // Empty the buffer
+                self.cap = 0;
+                self.pos = 0;
```
Comment (Diggsey) [inline src/libstd/io/buffered.rs]: Why do we need to mark the buffer as empty here, when we do the same further down?

Comment (Diggsey) [reply src/libstd/io/buffered.rs]: Never-mind just saw the conversation above!

**Comment by Diggsey** [comment]
What about `read_vectored` - I notice that doesn't even attempt to clear the buffer?

**Comment by Diggsey** [comment]
Also, it might be worth pulling this out into a method, `discard_buffer()` so that this mistake isn't repeated.

**Comment by Milack27** [comment]
@Diggsey Good point, I didn't understand yet how `read_vectored` works, so I didn't touch it. And I like the `discard_buffer()` idea, I'll do it.

**Comment by Diggsey** [comment]
It's pretty much the same as `read()`, just using multiple user-provided buffers instead of one. It needs a call to `discard_buffer()` in the same place as for the `read()` call.

**Comment by Milack27** [comment]
Does `discard_buffer` need any other annotation?

```Rust
   /// Invalidates all data in the internal buffer.
    #[inline]
    fn discard_buffer(&mut self) {
        self.pos = 0;
        self.cap = 0;
    }
```

**Comment by Milack27** [comment]
I had tried to improve the readability of the seek method, but I ended up changing its behavior as well, so I just rolled back.

**Comment by Dylan-DPC** [comment]
ping from triage @joshtriplett any updates on this?

**Pull request merged**